### PR TITLE
Add CalcHessian for H1-conforming tensor finite elements [h1-hessian-dev]

### PR DESCRIPTION
**Pull request by aschaf**
This PR adds an implementation of `CalcHessian` for `H1_SegmentElement`, `H1_QuadrilateralElement` and `H1_HexahedronElement`. So far, `Poly_1D::Basis::Eval(...)` has the second derivatives only implemented for `Poly_1D::EvalType::Barycentric`.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1474](https://github.com/mfem/mfem/pull/1474) | @aschaf | @tzanio | @v-dobrev + @stefanozampini | 05/19/20 | 08/13/20 | 08/20/20 | |

### CODE REVIEW

**Comment by tzanio** [comment]
@v-dobrev and @stefanozampini reminder that the reviews for this PR are due

**Comment by tzanio** [comment]
@v-dobrev and @stefanozampini -- do you need more time?

File: fem/fe.cpp [hunk 1]
```diff
@@ -7637,6 +7723,25 @@ void H1_SegmentElement::CalcDShape(const IntegrationPoint &ip,
...
+{
+   const int p = Order;
+
+#ifdef MFEM_THREAD_SAFE
+   Vector shape_x(p+1), dshape_x(p+1), d2shape_x(p+1);
+#endif
+
+   basis1d.Eval(ip.x, shape_x, dshape_x, d2shape_x);
+
+   Hessian(0,0) = d2shape_x(0);
```
Comment (stefanozampini) [inline fem/fe.cpp:7737]: why not just using a single loop?

File: fem/fe.cpp [hunk 2]
```diff
@@ -7849,6 +7982,34 @@ void H1_HexahedronElement::CalcDShape(const IntegrationPoint &ip,
...
+   Vector d2shape_x(p+1), d2shape_y(p+1), ds2hape_z(p+1);
+#endif
+
+   basis1d.Eval(ip.x, shape_x, dshape_x, d2shape_x);
+   basis1d.Eval(ip.y, shape_y, dshape_y, d2shape_y);
+   basis1d.Eval(ip.z, shape_z, dshape_z, d2shape_z);
+
+   for (int o = 0, k = 0; k <= p; k++)
+      for (int j = 0; j <= p; j++)
+         for (int i = 0; i <= p; i++)
```
Comment (stefanozampini) [inline fem/fe.cpp:8002]: please add o++ in the innermost loop increment. I missed it the first time I was reading the code

Comment (aschaf) [reply fem/fe.cpp:8002]: Should I also fix that in all occurences (e.g all `CalcDShape` methods)?

**Comment by aschaf** [comment]
> Is the new code tested?

So far I tested the hexahedral elements.

**Comment by tzanio** [comment]
@aschaf, can you resolve the conflicts and Travis failure?

@v-dobrev and @stefanozampini can you take another look at this PR?

**Comment by aschaf** [comment]
> @aschaf, can you resolve the conflicts and Travis failure?

Should I again merge 'master' to resolve the conflicts? The conflict is in an comment, which was changed by master and by this PR.

**Comment by tzanio** [comment]
> Should I again merge 'master' to resolve the conflicts?
Yes

**Comment by aschaf** [comment]
@tzanio all checks are now passing

**Comment by tzanio** [comment]
Thanks @aschaf.

@v-dobrev -- can you take a look?

**Comment by tzanio** [comment]
LGTM

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Pull request merged**